[export] construct empty graph when there's no tensor computation

yushangdi opened the pull request:
Fixes [#127110](https://github.com/pytorch/pytorch/issues/127110). 

When input module does not contain any tensor computation, we would create a graph with inputs and outputs.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

yushangdi committed:
construct empty graph

angelayi commented on torch/_dynamo/eval_frame.py:
```diff
@@ -1429,22 +1433,37 @@ def result_capturing_wrapper(*graph_inputs):
+            parameter_names = list(original_signature.parameters.keys())
+            fx_graph = torch.fx.Graph()
+            for i, name in enumerate(parameter_names):
+                if (torch.is_tensor(flat_args[i])):
```
We still want to create input nodes for constant inputs, not just tensors

yushangdi committed:
create input nodes for constant inputs

yushangdi committed:
fix typo

yushangdi committed:
add test cases

yushangdi committed:
lint

yushangdi committed:
lint

yushangdi committed:
add matching

yushangdi committed:
lint

yushangdi committed:
comment

angelayi commented on torch/_dynamo/eval_frame.py:
```diff
@@ -1429,22 +1431,55 @@ def result_capturing_wrapper(*graph_inputs):
+            for r in pytree.tree_flatten(result_traced)[0]:
+                if torch.is_tensor(r):
+                    graph_captured_result = graph_captured_result + (r,)
+            fx_graph.output(graph_captured_result)
```
this doesn't seem right? it still seems like we're directly inlining the tensor into the graph?

yushangdi replied:
Only tensors outputs are inlined here, and the output will be changed in rewrite_signature. For example, the output of the code below is
```
class Module(torch.nn.Module):
    def forward(self, x, y):
        return x

f = Module()
inp = ([torch.randn(3)], 1)
ep = torch.export.export(f, inp, None)
print(ep)
```

```
ExportedProgram:
    class GraphModule(torch.nn.Module):
        def forward(self, x_0: "f32[3]", y):
            return (x_0,)
            
Graph signature: ExportGraphSignature(input_specs=[InputSpec(kind=<InputKind.USER_INPUT: 1>, arg=TensorArgument(name='x_0'), target=None, persistent=None), InputSpec(kind=<InputKind.USER_INPUT: 1>, arg=ConstantArgument(name='y', value=1), target=None, persistent=None)], output_specs=[OutputSpec(kind=<OutputKind.USER_OUTPUT: 1>, arg=TensorArgument(name='x_0'), target=None)])
Range constraints: {}
```

Do we want to explicitly match and use the variables here before rewrite_signature?

angelayi replied:
Hmm, maybe I missed some changes when I was running it locally. My graph looked something like
```
def forward(self, x, y):
    return [tensor([ 0.3295, -0.2952,  0.3598])]
```

angelayi commented on test/export/test_export.py:
```diff
@@ -268,6 +268,42 @@ def forward(self, x, y):
+
+        f = Module()
+        inp = ([torch.randn(3)], 1)
+        self._test_export_same_as_eager(f, inp)
```
it would be nice to use [assertExpectedInline](https://github.com/pytorch/pytorch/blob/b019f38fdd97957f0e5923961ea7cf5b63c0761b/test/export/test_export.py#L4065-L4073) to check that the code matches what we expect

yushangdi replied:
added now

yushangdi committed:
add graph check

angelayi commented:
chatted offline -- also would be good to check the results of torch._dynamo.export and torch._dynamo.export(..., aten_graph=True)

yushangdi committed:
add dynamo export test, add error out

yushangdi commented:
> chatted offline -- also would be good to check the results of torch._dynamo.export and torch._dynamo.export(..., aten_graph=True)

added some tests for those. I also added the check for export.export() graph output. Do the output graphs look ok?

angelayi commented on test/dynamo/test_export.py:
```diff
@@ -607,6 +677,63 @@ def func():
+def forward(self, x, y):
+    arg0, arg1, = fx_pytree.tree_flatten_spec(([x, y], {}), self._in_spec)
+    arg0_1 = arg0
+    _tensor_constant0 = self._tensor_constant0
```
This is a little suspicious, but at least it's not being used in the result

yushangdi replied:
The _tensor_constant is created [here](https://github.com/pytorch/pytorch/blob/eeef68671d4f19d2f7d364ec6abb21d3c2696fad/torch/fx/_symbolic_trace.py#L822) when creating the aten graph. Is it reasonable?

yushangdi replied:
I made another minor modification to the code, we actually do not need to output anything for our graph, we just need to output ().

Since there's no tensor computation, any returned tensor can only come from the input. In that case, rewrite_signature will take care of adding it to the final graph output by looking at `result_traced`.

angelayi replied:
nice!

angelayi reviewed (approved):
looks great! thanks!

yushangdi committed:
fix output result

yushangdi committed:
remove output to graph

yushangdi commented:
@pytorchbot merge

yushangdi committed:
lint

yushangdi committed:
revert lint

yushangdi committed:
lint

yushangdi committed:
mark expectedFailureTrainingIRToRunDecomp in test

yushangdi commented:
@pytorchbot merge -i
